src/persist-client/src/operators/shard_source.rs

```diff
@@ -306,103 +311,110 @@ where
             }
         };
 
-        // Store the subscription handle in the shared slot so that it stays alive until both
-        // operators exit
-        let mut subscription = subscription_handle.borrow_mut();
-        let subscription =
-            subscription.insert(read.subscribe(as_of.clone()).await.unwrap_or_else(|e| {
-                panic!(
-                    "{}: {} cannot serve requested as_of {:?}: {:?}",
-                    name_owned, shard_id, as_of, e
-                )
-            }));
+        let snapshot_parts = if with_snapshot {
+            match read.snapshot(as_of.clone()).await {
+                Ok(parts) => futures::stream::iter(Some((parts, as_of.clone()))),
```

Comment:
Emitting the as_of as a progress after the snapshot here is, I think, a behavior change. Frank was recently mentioning that persist_source's behavior of emitting progress that is <= the as_of was causing some issues in arrangements, which led to [this workaround](https://github.com/MaterializeInc/materialize/blob/24cd2b53213a93ea3f9d313c1be7a125a3d8f7de/src/compute/src/render/mod.rs#L226-L231). So, while this progress is technically correct, I think it is not desirable. Seems easy enough to make the second part of this tuple an Option and fill in None here?

Reply:
Hm, a few lines above we are unconditionally downgrading the frontier of this operator to the `as_of`, which is why I thought it's safe to also put the `as_of` here, since it wouldn't change anything. Am I missing something?

Reply:
Ah, you're right. Github was hiding that from me in the diff and I was misremembering where that bit of code was. Disregard!